logprep/processor/processor_strategy.py

```diff
@@ -64,10 +71,17 @@ def _process_generic(
         processor_metrics: "Processor.ProcessorMetrics",
     ):
         """method for processing generic rules"""
-        for rule in generic_tree.get_matching_rules(event):
-            begin = time()
-            callback(event, rule)
-            processing_time = time() - begin
-            rule.metrics._number_of_matches += 1
-            rule.metrics.update_mean_processing_time(processing_time)
-            processor_metrics.update_mean_processing_time_per_event(processing_time)
+        applied_rules = set()
+        matching_rules = set(generic_tree.get_matching_rules(event))
+        while True:
```

Comment:
As I stated in my comment, I don't fully understand the logic inside `RuleTree` yet but independently from the functionality, is it really necessary to have the exact same line of codes a second time? I see the `tree` dependent looping, but this could be an own `loop` function which will get only the reference to each tree.

Reply:
I think the reason we did this is so we can explicitly test the call order of the two functions `process_specific` and `process_generic`. But you are right it is not really nice to have the code exactly twice. I could do something like this:

```python
def _process_specific(
    self,
    event: dict,
    specific_tree: "RuleTree",
    callback: Callable,
    processor_metrics: "Processor.ProcessorMetrics",
):
    self._process_rule_tree(event, specific_tree, callback, processor_metrics)

def _process_generic(
    self,
    event: dict,
    generic_tree: "RuleTree",
    callback: Callable,
    processor_metrics: "Processor.ProcessorMetrics",
):
    self._process_rule_tree(event, generic_tree, callback, processor_metrics)

def _process_rule_tree(
    self,
    event: dict,
    tree: "RuleTree",
    callback: Callable,
    processor_metrics: "Processor.ProcessorMetrics",
):
    # tree looping stuff comes here
```
That way we have minimized the redundant code and we still have two explicit methods we can test for.

Reply:
:+1: yes I prefer and like this much more as it increases readability also.